browser/modules/SearchProvider.jsm

```diff
@@ -0,0 +1,265 @@
+    let priv = PrivateBrowsingUtils.isBrowserPrivate(browser);
+    // fetch() rejects its promise if there's a pending request, but since we
+    // process our event queue serially, there's never a pending request.
+    let suggestions = yield controller.fetch(data.searchString, priv, engine);
+
+    if (!suggestions) {
+      throw new Error("Suggestions is null");
+    }
+
+    // Keep the form history result so RemoveFormHistoryEntry can remove entries
+    // from it.  Keeping only one result isn't foolproof because the client may
+    // try to remove an entry from one set of suggestions after it has requested
+    // more but before it's received them.  In that case, the entry may not
+    // appear in the new suggestions.  But that should happen rarely.
+    browserData.previousFormHistoryResult = suggestions.formHistoryResult;
+
+    let suggestion = {
+      engineName: data.engineName,
+      searchString: suggestions.term,
+      formHistory: suggestions.local,
+      remote: suggestions.remote,
+    };
+
+    return suggestion;
+  }),
+
+  addFormHistoryEntry: function(browser, entry) {
+    let isPrivate = true;
+    try {
+      // isBrowserPrivate assumes that the passed-in browser has all the normal
+      // properties, which won't be true if the browser has been destroyed.
+      // That may be the case here due to the asynchronous nature of messaging.
+      isPrivate = PrivateBrowsingUtils.isBrowserPrivate(browser);
+    } catch (err) {}
+    if (isPrivate || entry === "") {
+      return Promise.resolve();
+    }
+    let browserData = this._suggestionDataForBrowser(browser, true);
+    FormHistory.update({
+      op: "bump",
+      fieldname: browserData.controller.formHistoryParam,
+      value: entry,
+    }, {
+      handleCompletion: () => {},
+      handleError: err => {
+        Cu.reportError("Error adding form history entry: " + err);
+      },
+    });
+  },
+
+  removeFormHistoryEntry: function(browser, suggestionStr) {
+    let browserData = this._suggestionDataForBrowser(browser);
+    if (!browserData && !browserData.previousFormHistoryResult) {
+      return;
+    }
+    let {
+      previousFormHistoryResult
+    } = browserData;
+    for (let i = 0; i < previousFormHistoryResult.matchCount; i++) {
+      if (previousFormHistoryResult.getValueAt(i) === suggestionStr) {
+        previousFormHistoryResult.removeValueAt(i, true);
+        break;
+      }
+    }
+  },
+
+  _suggestionDataForBrowser: function(browser, create = false) {
+    let data = this._suggestionMap.get(browser);
+    if (!data && create) {
+      // Since one SearchSuggestionController instance is meant to be used per
+      // autocomplete widget, this means that we assume each xul:browser has at
+      // most one such widget.
+      data = {
+        controller: new SearchSuggestionController(),
+      };
+      this._suggestionMap.set(browser, data);
+    }
+    return data;
+  },
+
+  _arrayBufferFromDataURI: function(uri) {
```

Comment:
I know you didn't write this, but this seems like a total hack. Can we see of there is a different way to do this?